### DISCUSSION TITLE
New setup

### DISCUSSION THREAD

**Discussion by marcus-j-davies** [zwave-js/node-red-contrib-zwave-js#186] [Q&A]
@fabnavigator.

Is the serial port in use by zwavejs2mqtt?
if so - that's the reason. Serial ports can't be shared - only 1 application can obtain access.

But - yes, I believe you will need to expose the serial port in your container.
I am not sure how that is done - most users of this module are 'bare metal' Nod RED users - but it should all work the same never the less - given the extra setup needed for docker

**Comment by fabnavigator** [comment]
Docker is really cool, except for when there is an issue. The beauty is that the issue only affects a single container. Everything else works perfectly.

No it isn't I didn't spin up the zwave or  home-assistant containers.

If I don't add a device line to the docker-compose script for node-red, and try to deploy your device I get this:
```
24 Feb 14:10:02 - [error] [zwave-js:Z-Wave JS Controller] Error: Driver failed: Will retry in 5000ms. Attempted: 1, Max: 3
24 Feb 14:10:16 - [error] [zwave-js:Z-Wave JS Controller] Error: Driver failed: Will retry in 5000ms. Attempted: 2, Max: 3
24 Feb 14:10:30 - [error] [zwave-js:Z-Wave JS Controller] ZWaveError: Failed to open the serial port: Error: No such file or directory, cannot open /dev/ttyUSB0 (ZW0100)
```

That's what I would expect. Once I add the device line I get this:
```
24 Feb 14:06:06 - [error] [zwave-js:Z-Wave JS Controller] Error: Driver failed: Will retry in 5000ms. Attempted: 1, Max: 3
24 Feb 14:06:20 - [error] [zwave-js:Z-Wave JS Controller] Error: Driver failed: Will retry in 5000ms. Attempted: 2, Max: 3
24 Feb 14:06:34 - [error] [zwave-js:Z-Wave JS Controller] ZWaveError: Failed to open the serial port: Error: Permission denied, cannot open /dev/ttyUSB0 (ZW0100)
```
The docker user is in the dialout group, but just ro be safe I set the access like this:
```crw-rw-rw- 1 root dialout 188, 0 Feb 24 13:42 /dev/ttyUSB0```

**Reply by marcus-j-davies** [reply]
Ok try this.

`sudo usermod -a -G tty <user that node red/docker is running under>`
`sudo usermod -a -G dialout <user that node red/docker is running under>`

adding the docker user to both the dialout and tty group


you might have to spin up docker after

**Reply by fabnavigator** [reply]
I did that, and recreated the container. Node-red crashed when I deployed your node.
```
npm ERR! path /usr/src/node-red
npm ERR! command failed
npm ERR! signal SIGSEGV
npm ERR! command sh -c node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"
npm ERR! A complete log of this run can be found in:
npm ERR!     /data/.npm/_logs/2022-02-24T19_42_46_221Z-debug.log
```

**Reply by marcus-j-davies** [reply]
can you start Node RED in the terminal and see it start up ? - that error is a whole different mess. and should not be caused by the node.

it seems the error is at the very start.

**Reply by fabnavigator** [reply]
That error was after node-red came up normally, and then I added and deployed your node.
If your node is there when I start node-red, I get that error at startup and node-red crashes.

Could this have anything to do with the inside vs outside versions of npm? Inside the container I have 8.1.2 , and outside I have 8.5.1.

**Reply by marcus-j-davies** [reply]
Nah if its installed - NPM is no longer needed.
I do suggest you import this and run it however - given the version conflicts that may still be present.

I think this is all about `serialport` and the multiple node versions that are surrounding  it.

The below will build `serialport` using the node red container

```
[{"id":"e85dacfa92bb5af9","type":"exec","z":"2d7bf6e3.84c97a","command":"npm rebuild","addpay":"","append":"","useSpawn":"false","timer":"","winHide":false,"oldrc":false,"name":"","x":730,"y":720,"wires":[["c04da5e98fb4be3d"],["c04da5e98fb4be3d"],["c04da5e98fb4be3d"]]},{"id":"c04da5e98fb4be3d","type":"debug","z":"2d7bf6e3.84c97a","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":1010,"y":700,"wires":[]},{"id":"3e5a892aa42b96f9","type":"inject","z":"2d7bf6e3.84c97a","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":540,"y":720,"wires":[["e85dacfa92bb5af9"]]}]
```

**Reply by marcus-j-davies** [reply]
ohh - we might need to `cd` into your **/data** folder first

`cd /data && npm rebuild`

**Reply by marcus-j-davies** [reply]
If still no cigar - we can try one thing that takes docker out of the picture - I do think its related.

**Comment by fabnavigator** [comment]
That error is caused by the node. If the node is not there there are no errors at startup.  If I deploy your node I get an error. If I then restart the container with your node there I get this (from the command line):
```
node-red     | Welcome to Node-RED
node-red     | ===================
node-red     |
node-red     | 24 Feb 16:39:23 - [info] Node-RED version: v2.2.2
node-red     | 24 Feb 16:39:23 - [info] Node.js  version: v16.13.1
node-red     | 24 Feb 16:39:23 - [info] Linux 5.10.92-v8+ arm64 LE
node-red     | 24 Feb 16:39:24 - [info] Loading palette nodes
node-red     | 24 Feb 16:39:28 - [info] Dashboard version 3.1.5 started at /ui
node-red     | 24 Feb 16:39:28 - [info] Settings file  : /data/settings.js
node-red     | 24 Feb 16:39:28 - [info] Context store  : 'default' [module=localfilesystem]
node-red     | 24 Feb 16:39:28 - [info] Context store  : 'memory' [module=memory]
node-red     | 24 Feb 16:39:29 - [info] User directory : /data
node-red     | 24 Feb 16:39:29 - [warn] Projects disabled : editorTheme.projects.enabled=false
node-red     | 24 Feb 16:39:29 - [info] Flows file     : /data/flows.json
node-red     | 24 Feb 16:39:29 - [info] Server now running at http://127.0.0.1:1880/
node-red     | 24 Feb 16:39:29 - [info] Starting flows
node-red     | 24 Feb 16:39:29 - [info] [cronplus:9:00a 25th] createTask - index: 0, static: true, opt: {"name":"schedule1","topic":"topic1","payloadType":"default","payload":"","expressionType":"cron","expression":"0 0 9 25 * * *","location":"","offset":"0","solarType":"all","solarEvents":"sunrise,sunset"}
node-red     | 24 Feb 16:39:30 - [info] Started flows
node-red     | npm ERR! path /usr/src/node-red
node-red     | npm ERR! command failed
node-red     | npm ERR! signal SIGSEGV
node-red     | npm ERR! command sh -c node $NODE_OPTIONS node_modules/node-red/red.js $FLOWS "--userDir" "/data"
node-red     |
node-red     | npm ERR! A complete log of this run can be found in:
node-red     | npm ERR!     /data/.npm/_logs/2022-02-24T21_39_32_536Z-debug.log
node-red exited with code 1
```

**Reply by marcus-j-davies** [reply]
Ok, I am certain this is to do with `serialport` having issues with the environment (whether it's the node version, permissions or otherwise)

We know of a few users using it under docker with no adverse affects - but the environment  is all within.

I don't know your environment of course, but can you **_back it up_**, then clear it out - so you end up with a clean state in NR.
then install the module using the palette menu - this will then build the serial port package (providing the docker thing has all the tools) using the same environment that Node RED is using (of course if you have the energy still 😆 ) - the docker user will need write access to` /data` (as it will compile)

At the moment there are too many variables in place - but sudden crashes is not something we here of.

any reason you want to use docker? and not just install/use node red on the RPI OS its self? - not to change your set up of course - Im just curious

**Reply by fabnavigator** [reply]
It works! Installing from the palette menu was the ticket. I know that some time ago, I wasn't able to install nodes from the palette menu with docker. I always got an error, so I gave up. I've always installed them outside of docker but in the correct directory, and they have always worked. I will always use the palette menu now.

All I needed was to add this to my docker-compose.yaml file:
```
    devices:
      - /dev/ttyUSB0:/dev/ttyUSB0
```
Thank you!

On to the next step. I will try not to take up any more of your time (for the time being at least).

**Reply by marcus-j-davies** [reply]
> It works! Installing from the palette menu was the ticket.

excellent!  - that makes me believe it was an environmental factor with running the commands 'outside'

>  I will try not to take up any more of your time (for the time being at least).

Please don't hold back, if you need help, need questions answered - carry on. I am in the UK, so just account for the time difference, when waiting for replies 👍 

your devices will need to be interviewed  - so allow that to happen fully, you can see their status in the UI.

**Reply by fabnavigator** [reply]
@marcus-j-davies 

I wanted to share my final docker-compose.yaml file in case some more of your users want to try docker.  Docker is really cool, and if you think some of your users might be intersted I have some easy to use scripts that I could share.

```
version: '3.7'

services:

  node-red:
    container_name: node-red
    image: nodered/node-red:2.2.2-16   # Make sure to select a nodejs version 16 tag
    user: '<local-user-UID>' 
    ports:
      - 1880:1880
    volumes:
      - /home/<local-username>/opt/node-red:/data
      - /etc/localtime:/etc/localtime:ro
    environment:
      TZ: "America/New_York"  # Need to adjust for your local timezone
    devices:
      - /dev/ttyACM0:/dev/ttyACM0
    group_add:
      - dialout
```

**Comment by fabnavigator** [comment]
I've setup another Pi4 as a sandbox so I can play with your nodes w/o affecting my production system.  I have lots of spare z-wave hardware. 

When I click on the network map the dialog is too large for my laptop screen so the top section is cutoff. Is there somewhere I can change the window size?

**Reply by marcus-j-davies** [reply]
Hi @fabnavigator 

The default size of the map is; 1024 x 768 - you can resize it, by drag clicking one of the corners.
Admittedly, this does assume most resolutions are greater these days 😬 

We are working on 6.6.0 (almost there) which will see a review of the default map size as well as fresh paint on the UI in general.

**Reply by fabnavigator** [reply]
Hi @marcus-j-davies 

I've got an old laptop 1366 x 768, and chrome sucks up about 15% of the height. I don't know if you can size the window based on the current display settings or not. Possible save the size if the user resizes it, and use that next time it opens?

**Reply by marcus-j-davies** [reply]
Yeah, I'm sure I can do something - a few options here.
the new map will be coming in 6.6.0

Sneak Peek (Subject to change)

https://user-images.githubusercontent.com/55892693/155727792-e3466a49-60c8-4c74-b393-73d6ba7a38e3.mov

**Reply by fabnavigator** [reply]
I'm looking forward to it. Looks great.
